File: src/baobzi.f90
module baobzi
  use, intrinsic :: iso_c_binding
  implicit none

  type, bind(c) :: baobzi_input_t
    type(c_funptr) :: func = c_null_funptr
    type(c_ptr) :: data = c_null_ptr
    integer(c_int) :: dim = 0
    integer(c_int) :: order = 0
    real(c_double) :: tol = 0
    real(c_double) :: minimum_leaf_fraction = 0
    integer(c_int) :: split_multi_eval = 0
    integer(c_int) :: max_depth = 50
  end type baobzi_input_t

  interface
    function baobzi_init (input, center, half_length) bind(c) result(func)
      use, intrinsic :: iso_c_binding
      import :: baobzi_input_t
      type(baobzi_input_t), intent(in) :: input
      real(kind=c_double), dimension(*) :: center, half_length
      type(c_ptr) :: func
    end function baobzi_init

    subroutine baobzi_save (fin, fname) bind(c)
      use, intrinsic :: iso_c_binding
      type(c_ptr), intent(in), value :: fin
      character(kind=c_char), dimension(*), intent(in) :: fname
      type(c_ptr) :: func
    end subroutine baobzi_save

    function baobzi_restore (fname) bind(c) result(func)
      use, intrinsic :: iso_c_binding
      character(kind=c_char), dimension(*), intent(in) :: fname
      type(c_ptr) :: func
    end function baobzi_restore

    function baobzi_eval (f, x) bind(c) result(y)
      use, intrinsic :: iso_c_binding
      type(c_ptr), intent(in), value :: f
      real(kind=c_double), dimension(*) :: x
      real(kind=c_double) :: y
    end function baobzi_eval

    subroutine baobzi_eval_multi (f, x, res, ntrg) bind(c)
      use, intrinsic :: iso_c_binding
      type(c_ptr), intent(in), value :: f
      real(kind=c_double), dimension(*) :: x
      real(kind=c_double), dimension(*) :: res
      integer(kind=c_int), value :: ntrg
    end subroutine baobzi_eval_multi

    subroutine baobzi_stats (fin) bind(c)
      use, intrinsic :: iso_c_binding
      type(c_ptr), intent(in), value :: fin
      type(c_ptr) :: func
    end subroutine baobzi_stats

    function baobzi_free (f) bind(c) result(f_null)
      use, intrinsic :: iso_c_binding
      type(c_ptr), intent(in), value :: f
      type(c_ptr) :: f_null
    end function baobzi_free
  end interface

end module baobzi

File: src/julia/baobzi.jl
module baobzi

mutable struct baobzi_t
end

mutable struct baobzi_input_t
    func::Ptr{Cvoid}
    data::Ptr{Cvoid}
    dim::Cint
    order::Cint
    tol::Cdouble
    minimum_leaf_fraction::Cdouble
    split_multi_eval::Cint
end

function init(fin, dim, order, center, half_length, tol, minimum_leaf_fraction=0.0, split_multi_eval=1)
    fanon = (x, p) -> fin(x)
    fbind = @cfunction($fanon, Cdouble, (Ptr{Cdouble}, Ptr{Cvoid},))
    input = baobzi_input_t(fbind.ptr, C_NULL, dim, order, tol, minimum_leaf_fraction, split_multi_eval)

    output_ptr = ccall(
        (:baobzi_init, :libbaobzi),
        Ptr{baobzi_t},
        (Ref{baobzi_input_t}, Ptr{Cdouble}, Ptr{Cdouble},),
        input, center, half_length
    )

    return output_ptr
end

function eval(func::Ptr{baobzi_t}, x)
    return ccall(
        (:baobzi_eval, :libbaobzi),
        Cdouble,
        (Ptr{Cvoid}, Ptr{Cdouble},),
        func, x
    )
end

function eval_multi(func::Ptr{baobzi_t}, x)
    n_points = Int32(length(x) // 2)
    println(n_points)
    res = Array{Float64}(undef, n_points)
    ccall(
        (:baobzi_eval_multi, :libbaobzi),
        Cvoid,
        (Ptr{Cvoid}, Ptr{Cdouble}, Ptr{Cdouble}, Cint,),
        func, x, res, n_points
    )
    return res
end

function stats(func::Ptr{baobzi_t})
    ccall(
        (:baobzi_stats, :libbaobzi),
        Ptr{Cvoid},
        (Ptr{Cvoid},),
        func
    )
end

function free(func::Ptr{baobzi_t})
    ccall(
        (:baobzi_free, :libbaobzi),
        Ptr{Cvoid},
        (Ptr{Cvoid},),
        func
    )
end

function save(func::Ptr{baobzi_t}, filename::String)
    ccall(
        (:baobzi_save, :libbaobzi),
        Ptr{Cvoid},
        (Ptr{Cvoid}, Cstring),
        func, filename
    )
end

function restore(filename::String)
    output_ptr = ccall(
        (:baobzi_restore, :libbaobzi),
        Ptr{baobzi_t},
        (Cstring,),
        filename
    )

    return output_ptr
end

end
